src/java.base/share/classes/java/lang/ref/package-info.java

```diff
@@ -93,6 +93,41 @@
  * structure, this check will add little overhead to the hashtable
  * access methods.
  *
+ * <a id="MemoryConsistency"></a>
+ * <h3>Memory Consistency Properties</h3>
+ * Certain interactions between the garbage collector, references, and reference
+ * queues form
+ * <a href="{@docRoot}/java.base/java/util/concurrent/package-summary.html#MemoryVisibility"><i>happens-before</i></a>
+ * relationships:
+ *
+ * <ul>
+ *
+ * <li> Actions in a thread prior to calling
+ * {@link Reference#reachabilityFence Reference.reachabilityFence(x)}
+ * <i>happen-before</i> the garbage collector clears any reference to {@code x}.</li>
+ *
+ * <li>The clearing of a reference by the garbage collector <i>happens-before</i>
+ * the garbage collector enqueues the reference.</li>
+ *
+ * <li> The enqueueing of a reference (by the garbage collector, or
+ * by a successful call to {@link Reference#enqueue}) <i>happens-before</i>
+ * the reference is removed from the queue ("dequeued") by {@link ReferenceQueue#poll} or
+ * {@link ReferenceQueue#remove}.</li>
+ *
+ * <li>The Cleaner thread dequeues a reference to a registered object before
+ * running the cleaning action for that object.</li>
```

Comment:
Do you want to rephrase this so that it too mentions `happens-before`? e.g.
```
Dequeuing of a reference to a registered object, by the Cleaner thread, happens-before the execution of the cleaning action for that object.
```

Reply:
I hesitated to state it that way because there's not a synchronization action _per se_ between the Cleaning thread dequeuing, and then running the cleaning action.

However, I see that JLS 17.4.5 does state:
_"If x and y are actions of the same thread and x comes before y in program order, then hb(x, y)."_
(That is, x _happens-before_ y).

So yes, we can say _happens-before_ here. Thanks!